### PR TITLE
Firmware Signing Post

### PR DESCRIPTION
**Pull request by franc0is**

### CODE REVIEW

File: _drafts/fwup-signing.md [hunk 1]
```diff
@@ -0,0 +1,642 @@
...
+also known as a secret key, but can be verified as genuine by anyone.
+
+Additionally, the signatures are tied to a given firmware binary so should the
+binary change by so much as 1 bit the signature would no longer be valid.
+
+### Why sign our firmware
+
+By implementing signature verification in our bootloader we can identify
+whether or not a given firmware update was provided by the manufacturer, or if
+it has been tempered with. The bootloader can then decide to either warn the
```
Comment (tyhoff) [inline _drafts/fwup-signing.md:37]:
```suggestion
it has been tampered with. The bootloader can then decide to either warn the
```

File: _drafts/fwup-signing.md [hunk 2]
```diff
@@ -0,0 +1,642 @@
...
+user, void the device's warranty, or simply refuse to run the unauthenticated
+binary.
+
+With more and more devices connected to the internet, security is an
+increasingly hot topic in firmware development. A device which accepts firmware
+updates over the wireless or internet connectivity but does not verify it opens
+itself to compromise. By feeding it with a malicious firmware image, attacker
```
Comment (tyhoff) [inline _drafts/fwup-signing.md:44]:
```suggestion
itself to compromise. By feeding it with a malicious firmware image, an attacker
```

File: _drafts/fwup-signing.md [hunk 3]
```diff
@@ -0,0 +1,642 @@
...
+might:
+* Brick the device, or the whole fleet
+* Snoop on end users and compromise their privacy and security
+* Strategically malfunction at a critical time
+
+These are highly undesirable outcome, which can be effected at scale due to the
```
Comment (tyhoff) [inline _drafts/fwup-signing.md:50]:
```suggestion
These are highly undesirable outcomes, which can be effected at scale due to the
```

File: _drafts/fwup-signing.md [hunk 4]
```diff
@@ -0,0 +1,642 @@
...
+> security, it is not by itself sufficient to build secure systems. Secure
+> coding, static analysis, hardware tamper detection, JTAG locking, and many
+> more techniques should be implemented as well.
+> Code signing also does nothing to protect against reverse engineering. It is a
+> distinct technique from firmware encryption.
+
+### ECDSA
+
+Several algorithms can be used to sign firmware, including RSA, DSA, and ECDSA.
+In this post, we focus on ECDSA for a few reason:
```
Comment (tyhoff) [inline _drafts/fwup-signing.md:64]:
```suggestion
In this post, we focus on ECDSA for a few reasons:
```

File: _drafts/fwup-signing.md [hunk 5]
```diff
@@ -0,0 +1,642 @@
...
+   using matching inputs.
+
+## Firmware Signing Implementation
+
+### Setup
+
+Like we did in previous post, we use [Renode]({% post_url
+2020-03-23-intro-to-renode %}) to run the examples in this post. Our previous
+[firmware update post]({% post_url 2020-06-23-device-firmware-update-cookbook
+%}) contains detailed instructions. In short:
```
Comment (tyhoff) [inline _drafts/fwup-signing.md:102]:
```suggestion
Like we did our previous post, the [Firmware Update Cookbook]({% post_url 2020-06-23-device-firmware-update-cookbook%}), we use [Renode]({% post_url
2020-03-23-intro-to-renode %}) to run the examples in this post. The previous post contains detailed instructions, but in short:
```

File: _drafts/fwup-signing.md [hunk 6]
```diff
@@ -0,0 +1,642 @@
...
+```
+
+One change we did make from previous posts is run in headless mode. This is done
+by invoking Renode with the following flags:
+
+```
+mono64 $RENODE_EXE_PATH renode-config.resc --port 4444 --disable-xwt
+```
+
+We can then use telnet to connect to the renode console:
```
Comment (tyhoff) [inline _drafts/fwup-signing.md:120]:
```suggestion
We can then use telnet to connect to the Renode console:
```

File: _drafts/fwup-signing.md [hunk 7]
```diff
@@ -0,0 +1,642 @@
...
+1. Bootloader: a simple program whose sole job is to load the Loader, and fallback
+   to another image if the Loader is invalid.
+2. Loader: a program that can verify our Application image, load it, and update
+   it.
+3. Application: our main code, which does not do any updates itself
+4. Updater: a program that temporarily replaces the Application and can update
+   the Loader.
+
+The Loader, the Application, and the Updater all need to be signed since they
+can be updated. Components that updates other components need to implement our
```
Comment (tyhoff) [inline _drafts/fwup-signing.md:216]:
```suggestion
can be updated. Components that update other components need to implement our
```

File: _drafts/fwup-signing.md [hunk 8]
```diff
@@ -0,0 +1,642 @@
...
+signature verification algorithm, in our case that's the Loader and the Updater.
+
+### Generating public/private key pairs
+
+Before we can begin, we must generate a pair of keys. Our **private key** also
+known as **signing key** is used to create the signatures, and should be kept
```
Comment (tyhoff) [inline _drafts/fwup-signing.md:222]:
```suggestion
Before we can begin, we must generate a pair of keys. Our **private key**, also
known as a **signing key**, is used to create the signatures, and should be kept
```

File: _drafts/fwup-signing.md [hunk 9]
```diff
@@ -0,0 +1,642 @@
...
+private. Anyone with access to the key will be able to sign firmware on your
+behalf. Our **public key** also known as **validation key** is used to verify
+signature. It can be freely distributed.
```
Comment (tyhoff) [inline _drafts/fwup-signing.md:225]:
```suggestion
behalf. Our **public key**, also known as a **validation key**, is used to verify
the signature. It can be freely distributed.
```

File: _drafts/fwup-signing.md [hunk 10]
```diff
@@ -0,0 +1,642 @@
...
+
+Several tools can be used to generate our key pair but the simplest is
+`openssl`, a cross platform cryptography tools.
```
Comment (tyhoff) [inline _drafts/fwup-signing.md:228]:
```suggestion
`openssl`, a cross platform cryptography toolset.
```

File: _drafts/fwup-signing.md [hunk 11]
```diff
@@ -0,0 +1,642 @@
...
+```
+
+Where `PK_PEM_PATH` is the path to your private key.<!-- TODO: Add a note about
+security -->
+
+We now have a build process that generates signed binaries!
+
+### Verifying signatures in our Loader
+
+Now that our firmware builds are signed, lets verify those signatures in our
```
Comment (tyhoff) [inline _drafts/fwup-signing.md:392]:
```suggestion
Now that our firmware builds are signed, let's verify those signatures in our
```

File: _drafts/fwup-signing.md [hunk 12]
```diff
@@ -0,0 +1,642 @@
...
+}
+```
+
+Note that we had to select the same curve we used to generate the key, here
+`secp2561k`.
+
+Last but not least, we need to verify the signature itself. This requires
+computing the SHA-256, and using `uECC_verify`.
+
+Many MCUs have hardware implementations of SHA-256, but in our case we used a
```
Comment (tyhoff) [inline _drafts/fwup-signing.md:465]:
```suggestion
Many MCUs have hardware implementations of SHA-256, but in our case, we used a
```

File: _drafts/fwup-signing.md [hunk 13]
```diff
@@ -0,0 +1,642 @@
...
+static void prv_sha256(const void *buf, uint32_t size, uint8_t *hash_out)
+{
+  cf_sha256_context ctx;
+  cf_sha256_init(&ctx);
+  cf_sha256_update(&ctx, buf, size);
+  cf_sha256_digest_final(&ctx, hash_out);
+}
+```
+
+Note that `cf_sha256_update` can be called mutliple time, so if you are
```
Comment (tyhoff) [inline _drafts/fwup-signing.md:483]:
```suggestion
Note that `cf_sha256_update` can be called mutliple times, so if you are
```

File: _drafts/fwup-signing.md [hunk 14]
```diff
@@ -0,0 +1,642 @@
...
+
+<!-- excerpt start -->
+<!-- excerpt end -->
+
+{% include newsletter.html %}
+{% include toc.html %}
+
+## Firmware Signing Explained
+
+### What is code signing
```
Comment (tyhoff) [inline _drafts/fwup-signing.md:15]: You can probably remove this specific header

File: _drafts/fwup-signing.md [hunk 15]
```diff
@@ -0,0 +1,639 @@
...
+bottom of our credit card receipts, contracts, and bank checks. These signatures
+can be verified quite easily by comparing them to previous signatures, but are
+hard to forge because they rely on an individual's muscle memory.
+
+Code signing uses cryptographic algorithms to achieve a similar goal. The
+resulting signatures can only be generated by people who know a given password,
+also known as a secret key, but can be verified as genuine by anyone.
+
+Additionally, the signatures are tied to a given firmware binary so should the
+binary change by so much as 1 bit the signature would no longer be valid.
```
Comment (tyhoff) [inline _drafts/fwup-signing.md:31]: I'm struggling with words today, but you jump right into firmware binaries without yet talking about firmware binaries. I think I'm missing how this all applies to firmware from a high level.

Something like:

When a signature is generated for a firmware binary, it is uniquely tied to that exact firmware. Changing the binary so much as 1 bit would produce a different signature.

With this in mind, if a device is given a signed binary and a signature and can verify both that the binary and signature are valid and that the binary was created by the correct author, then we know the binary has not been tampered with.


Basically, connect the dots for me, either in a diagram or words, before we get into this stuff where we are dealing directly with openssl

File: _drafts/fwup-signing.md [hunk 16]
```diff
@@ -0,0 +1,639 @@
...
+
+### Why sign our firmware
+
+By implementing signature verification in our bootloader we can identify
+whether or not a given firmware update was provided by the manufacturer, or if
+it has been tampered with. The bootloader can then decide to either warn the
+user, void the device's warranty, or simply refuse to run the unauthenticated
+binary.
+
+With more and more devices connected to the internet, security is an
```
Comment (tyhoff) [inline _drafts/fwup-signing.md:41]:
```suggestion
With more and more devices connected to the Internet, security is an
```

File: _drafts/fwup-signing.md [hunk 17]
```diff
@@ -0,0 +1,639 @@
...
+increasingly hot topic in firmware development. A device which accepts firmware
+updates over the wireless or internet connectivity but does not verify it opens
```
Comment (tyhoff) [inline _drafts/fwup-signing.md:43]:
```suggestion
updates wirelessly or over Internet connectivity but does not verify it opens
```

File: _drafts/fwup-signing.md [hunk 18]
```diff
@@ -0,0 +1,639 @@
...
+itself to compromise. By feeding it with a malicious firmware image, an attacker
+might:
+* Brick the device, or the whole fleet
+* Snoop on end users and compromise their privacy and security
+* Strategically malfunction at a critical time
```
Comment (tyhoff) [inline _drafts/fwup-signing.md:48]: If you want to open the router can of worms...create a botnet of Linux routers for funsies.

File: _drafts/fwup-signing.md [hunk 19]
```diff
@@ -0,0 +1,639 @@
...
+
+These are highly undesirable outcomes, which can be effected at scale due to the
+internet of things. In 2020, it is reckless to implement firmware update for our
```
Comment (tyhoff) [inline _drafts/fwup-signing.md:51]:
```suggestion
Internet of things. In 2020, it is reckless to implement firmware update for our
```

File: _drafts/fwup-signing.md [hunk 20]
```diff
@@ -0,0 +1,639 @@
...
+systems without some form of authentication.
+
+> **What signing is not**: code signing is an important component of firmware
```
Comment (tyhoff) [inline _drafts/fwup-signing.md:54]: Maybe even move it to its own section rather than a big block comment.

File: _drafts/fwup-signing.md [hunk 21]
```diff
@@ -0,0 +1,639 @@
...
+> security, it is not by itself sufficient to build secure systems. Secure
+> coding, static analysis, hardware tamper detection, JTAG locking, and many
+> more techniques should be implemented as well.
+> Code signing also does nothing to protect against reverse engineering. It is a
+> distinct technique from firmware encryption.
+
+### ECDSA
```
Comment (tyhoff) [inline _drafts/fwup-signing.md:61]: I think this section is quite dry, and less exciting than FIrmware Signing implementation. 

Maybe you can quickly say `We use ECDSA, which we talk about [later in the post](#ecdsa)` and jump straight to the example, and place this section later.

**Review by tyhoff — COMMENTED** [review]
Key Storage and maybe Developer Keys could be touched upon by our guest post in review. Maybe reserve those two? Or at least key storage.

EDIT: Ohh, private key storage. Got it. Yeah, great topic to cover.

File: _drafts/fwup-signing.md (hunk 14 above)
Comment (franc0is) [reply _drafts/fwup-signing.md:34]: :+1:

File: _drafts/fwup-signing.md (hunk 15 above)
Comment (franc0is) [reply _drafts/fwup-signing.md:50]: Used your words, they're great

File: _drafts/fwup-signing.md (hunk 18 above)
Comment (franc0is) [reply _drafts/fwup-signing.md:67]: I decided not to call anyone out explicitly.

File: _drafts/fwup-signing.md (hunk 20 above)
Comment (franc0is) [reply _drafts/fwup-signing.md:73]: I don't think I have quite enough for a section

File: _drafts/fwup-signing.md (hunk 21 above)
Comment (franc0is) [reply _drafts/fwup-signing.md:80]: I added a "jump to implementation" at the top of the theory section. At the same time, I think this needs to be before the implementation, because it details the use of SHA-256 which is key to the implementation.

**Comment by tyhoff** [comment]
You've got some work to do on the links, but other than that, I like it! Love the openssl example usage and examples.

For the title, pick one (I think):

Secure firmware update[s] with code signing
Secur[ing] firmware update with code signing

**Comment by tyhoff** [comment]
Maybe change the slug of the post to something more descriptive and SEO friendly? fwup-signing is pretty cryptic.

**Review by tyhoff — APPROVED** [review]
